src/main/scala/karazin/scala/users/group/week5/homework/givens.scala

```diff
@@ -1,19 +1,46 @@
 package karazin.scala.users.group.week5.homework
 
+import scala.annotation.tailrec
+
 object givens:
-  
+
   /* 
     The trait is used for converting instances to a json string representation
     Provide a type parameter(s) for the trait and the method 
     and argument(s) and a result type
-  */ 
-  
-  trait DummyType
-  
-  trait JsonStringEncoder:
-    def encode: DummyType
+  */
 
-  /* 
-    Make sure that integers, booleans, strings and lists 
-    are convertable to a json string representation 
-  */ 
\ No newline at end of file
+
+  trait JsonEncoder[T]:
+    def encode(element: T): String
+
+  given StringEncoder: JsonEncoder[String] with
+    def encode(element: String): String =
+      '"' + element + '"'
+
+
+  given IntEncoder: JsonEncoder[Int] with
+    def encode(element: Int): String =
+      element.toString
+
+
+  given BooleanEncoder: JsonEncoder[Boolean] with
+    def encode(element: Boolean): String =
+      if element then "true"
+      else "false"
+
+  given ListEncoder[T] (using encoder: JsonEncoder[T]): JsonEncoder[List[T]] with
+    def encode(element_List: List[T]): String =
+      @tailrec
+      def encode_elements(elementList: List[T], pos: Int, ans: String): String =
```

Comment:
Oh, my.
Looks like I didn't get your question about the recursion.
I'd like to see here the following solution:
1. use `map` (https://www.scala-lang.org/api/2.13.3/scala/collection/immutable/List.html#map[B](f:A=%3EB):List[B]) to encode all elements;
2. use `mkString` (https://www.scala-lang.org/api/2.13.3/scala/collection/immutable/List.html#mkString(start:String,sep:String,end:String):String) to build a string from the list.

Reply:
can we use here flatmap?